### PR TITLE
WIP FB8-65: Add backward compatibility for "COMPRESSION=" options

### PR DESCRIPTION
**Pull request by inikep**
Jira issue: https://jira.percona.com/browse/FB8-65

Reference Patch: https://github.com/facebook/mysql-5.6/commit/d4057fa

Summary:
Since table DDLs dumped/shown from 5.6.24 may still contain "COMPRESSION="
options that are dropped in 5.6.27, they will fail to load directly into
5.6.27. I am adding the syntax backward compatibility back, and ignoring the
options in the DDL.

The options no longer supported and ignored from 5.6.24 are: COMPRESSION,
COMPRESSION_LEVEL and COMPACT_METADATA.

Once all systems are upgraded to 5.6.27, we should no longer see those options
in the DDLs from mysqldump or SHOW CREATE TABLE.

Test Plan:
Added back innodb.innodb_create_table_options. Make sure the tables with those
options are created successfully, while `SHOW CREATE TABLE` will NOT have those
options in the DDL.

Reviewers: hermanlee4, pengt

Originally Reviewed By: pengt

Subscribers: domas, webscalesql-eng, ebergen

### CODE REVIEW

File: mysql-test/suite/innodb/t/innodb_create_table_options.test [hunk 1]
```diff
@@ -0,0 +1,285 @@
+--source include/have_innodb_max_16k.inc
+let $file_per_table_save = `SELECT @@innodb_file_per_table`;
+SET GLOBAL innodb_file_per_table=ON;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/innodb_create_table_options.test]: It's the default already

File: mysql-test/suite/innodb/t/innodb_create_table_options.test [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+CREATE TABLE t_snappy2(a INT PRIMARY KEY) ENGINE=InnoDB ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=4 COMPRESSION=SNAPPY COMPACT_METADATA=1;
+CREATE TABLE t_quicklz1(a INT PRIMARY KEY) ENGINE=InnoDB ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8 COMPRESSION=QUICKLZ COMPRESSION_LEVEL=2;
+CREATE TABLE t_quicklz2(a INT PRIMARY KEY) ENGINE=InnoDB ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=4 COMPRESSION=QUICKLZ COMPRESSION_LEVEL=3 COMPACT_METADATA=1;
+CREATE TABLE t_lz41(a INT PRIMARY KEY) ENGINE=InnoDB ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8 COMPRESSION=LZ4 COMPRESSION_LEVEL=2;
+CREATE TABLE t_lz42(a INT PRIMARY KEY) ENGINE=InnoDB ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=4 COMPRESSION=LZ4 COMPRESSION_LEVEL=0 COMPACT_METADATA=1;
+
+
+# restart mysql to make sure that the table schemata are persisted.
+--source include/restart_mysqld.inc
+SET GLOBAL innodb_file_per_table=ON;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/innodb_create_table_options.test]: Likewise

File: sql/parse_tree_nodes.h [hunk 3]
```diff
@@ -3231,6 +3231,18 @@ class PT_create_table_default_collation : public PT_create_table_option {
   bool contextualize(Table_ddl_parse_context *pc) override;
 };
 
+class PT_ignore_create_table_option : public PT_create_table_option {
+  typedef PT_create_table_option super;
+
+ public:
+  explicit PT_ignore_create_table_option() {}
```
Comment (laurynas-biveinis) [inline sql/parse_tree_nodes.h]: No need to mark zero-arg constructor as explicit. Should it call parent default constructor, if any? Can it be marked noexcept while we are at that?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: sql/parse_tree_nodes.h (hunk 3 above)
Comment (inikep) [reply sql/parse_tree_nodes.h]: I removed the constructor.

**Comment by inikep** [comment]
PR was updated.

File: sql/sql_yacc.yy [hunk 4]
```diff
@@ -5926,9 +5935,21 @@ create_table_option:
             $$= NEW_PTN PT_create_commen_option($3);
           }
         | COMPRESSION_SYM opt_equal TEXT_STRING_sys
-	  {
+          {
             $$= NEW_PTN PT_create_compress_option($3);
-	  }
+          }
+        | COMPRESSION_SYM opt_equal compression_types
```
Comment (percona-ysorokin) [inline sql/sql_yacc.yy:5941]: @hermanlee 
Is there a collision (at least logical) between InnoDB transparent page compression
`| COMPRESSION_SYM opt_equal TEXT_STRING_sys` (for instance `COMPRESSION='ZLIB'`, notice quotes here), which is a meaningful directive
and
`| COMPRESSION_SYM opt_equal compression_types` (for instance `COMPRESSION=ZLIB`, notice no quotes here), which should be ignored.
In my opinion, this is a bit confusing for the end user

File: sql/parse_tree_nodes.h [hunk 5]
```diff
@@ -3231,6 +3231,16 @@ class PT_create_table_default_collation : public PT_create_table_option {
   bool contextualize(Table_ddl_parse_context *pc) override;
 };
 
+class PT_ignore_create_table_option : public PT_create_table_option {
+  typedef PT_create_table_option super;
```
Comment (percona-ysorokin) [inline sql/parse_tree_nodes.h]:
```suggestion
  typedef using super = PT_create_table_option;
```

File: mysql-test/suite/innodb/t/innodb_create_table_options.test [hunk 6]
```diff
@@ -0,0 +1,281 @@
+--source include/have_innodb_max_16k.inc
+
+############################# ROW_FORMAT=COMPACT ###############################
+# Make sure that compact table creation works as usual.
+CREATE TABLE t_compact(a INT PRIMARY KEY) ENGINE=InnoDB ROW_FORMAT=COMPACT;
+SHOW CREATE TABLE t_compact;
+DROP TABLE t_compact;
+# Make sure that creating an compact table with COMPRESSION parameter set issues a warning and ignores this parameter
```
Comment (percona-ysorokin) [inline mysql-test/suite/innodb/t/innodb_create_table_options.test:8]: Here and everywhere, no warning is issued in the result file.

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: mysql-test/suite/innodb/t/innodb_create_table_options.test (hunk 6 above)
Comment (inikep) [reply mysql-test/suite/innodb/t/innodb_create_table_options.test:8]: It's the same in the original patch.

Comment (inikep) [reply mysql-test/suite/innodb/t/innodb_create_table_options.test:8]: Should I remove these comments or add a warning?

Comment (percona-ysorokin) [reply mysql-test/suite/innodb/t/innodb_create_table_options.test:8]: @hermanlee ^^

**Pull request closed without merging**